### ISSUE TITLE
0.4b: Error during setup of component indego

### ISSUE THREAD

**Issue by onkelfarmor** [jm-73/Indego#6]
Get this error after copying the new files to my folder and restarting HA.

```
Error during setup of component indego
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/setup.py", line 156, in _async_setup_component
    component.setup, hass, processed_config)  # type: ignore
  File "/usr/local/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/config/custom_components/indego/__init__.py", line 68, in setup
    IndegoAPI_Instance = IndegoAPI(username=mower_username, password=mower_password, serial=mower_serial)
  File "/config/custom_components/indego/__init__.py", line 177, in __init__
    self.update()
  File "/config/custom_components/indego/__init__.py", line 318, in update
    self.getGenericData()
  File "/config/custom_components/indego/__init__.py", line 526, in getGenericData
    self._alm_name = tmp_json['alm_name']
KeyError: 'alm_name'
```

**Comment by jm-73** [comment]
Are there any other info about the component before this rows in your log?

**Comment by onkelfarmor** [comment]
nope :-(

**Comment by jm-73** [comment]
If you do another reboot? If you try to delete the __pycache__ in your Indego folder before reboot?

**Comment by jm-73** [comment]
I have tested with a fresh install of hassio 0.96.5 right now. Cant reproduce your error. 
Also rnning without error messages in my development environment as well as my "real" Hassio.
So I need more infomration from you in order to solve this.

**Comment by onkelfarmor** [comment]
Tried everything.
- rebooted several times.
- removed __pycache__ and rebooted several times
- removed the whole package, rebooted several times and installed again
- installed the old version (0.3) - this works.
- tried reinstalling the beta version - same error

Using hassio 0.65.5

**Comment by onkelfarmor** [comment]
I forgot to write, that I also get this error in the frontend:

```
The following components and platforms could not be set up:

indego
Please check your config.
```

But I haven't changed my config, so it looks like it now expects something ells from the config - don't see any changes about that in the instruction.

**Comment by nsimb** [comment]
also got error setting it up after the update. Let me know if you need anything

**Comment by jm-73** [comment]
Hmmm, very strange. The only thing I need is the log file. But if it says nothing more than you stated here, then it is not of any use.

@nsimb : What version of Hassio are you running? Or are you running HA?

**Comment by nsimb** [comment]
Home Assistant 0.96.5
Hass.io 170
HassOS 2.12

Reverted to the previous version and it works again, sadly lost the log file.
I can redo the change and get you a new logfile if you need mine

**Comment by jm-73** [comment]
The interesting part is that you got different kind of errors. The error indicates that the json answer doesnt contain the expected values.
I think I have to add some error handling as my component handles the json answers very static.

**Comment by jm-73** [comment]
You have tried the debug log?

``` #configuration.yaml
logger:
  logs:
    custom_components.indego: debug
```

**Comment by jm-73** [comment]
If you are comfortable with changing in your component code, please change the following:

@onkelfarmor 
`__init__.py` row 526:
```self._alm_name = tmp_json['alm_name']```
to
```self._alm_name = tmp_json.get('alm_name')```

@nsimb 
`__init__.py` row 507:
```self._optinapp = tmp_json['optInApp']```
to
```self._optinapp = tmp_json.get('optInApp')```

If you could try this and see if the error moves to another row, that would be of great help to me. If this json string handling is fixing the error, then I could release a new beta that is corrected.

**Comment by onkelfarmor** [comment]
I changed row 526 as described and now get a new error, so yes, it moves to a another row:

```
Error during setup of component indego
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/setup.py", line 156, in _async_setup_component
    component.setup, hass, processed_config)  # type: ignore
  File "/usr/local/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/config/custom_components/indego/__init__.py", line 68, in setup
    IndegoAPI_Instance = IndegoAPI(username=mower_username, password=mower_password, serial=mower_serial)
  File "/config/custom_components/indego/__init__.py", line 177, in __init__
    self.update()
  File "/config/custom_components/indego/__init__.py", line 319, in update
    self.ModelDescription()
  File "/config/custom_components/indego/__init__.py", line 826, in ModelDescription
    return self._model_description
AttributeError: 'IndegoAPI' object has no attribute '_model_description'
```

**Comment by jm-73** [comment]
Ok, good. My theory works then!

You could try to continue changing all the locations that gives you an error. Or you could wait for an updated beta from me with this error control in all json handling.

**Comment by onkelfarmor** [comment]
Would love to do it myself, but need more guidance :-)
or I just wait for you. 

Very happy about your work!

**Comment by jm-73** [comment]
There is another beta tagged as 0.4b2. IT is only the `__init__.py'` that is updated with the new json handling. So it should be enough to just replace this file, restart and look in the log file...

**Comment by onkelfarmor** [comment]
Updated, I now don't get any errors but don't know how to at what to look for in the log.

**Comment by jm-73** [comment]
See if your sensors are working. Other than that, look for errors in the log.
Are your sensors showing as expected?

**Comment by onkelfarmor** [comment]
After some restarts and some retries, I now get the same warning, that Indego not could be setup and the same error in the log as above. No sensors working.

**Comment by onkelfarmor** [comment]
Found an error in __init.py__, line 811:

`return self._model_description`

must be

` return self.model_description`

now I get more sensors showing up, but they are not updated with data. However, in the log I see that all data is collected from Bosch, but it is somehow not written to the sensor.

**Comment by nsimb** [comment]
> If you are comfortable with changing in your component code, please change the following:
> 
> @onkelfarmor
> `__init__.py` row 526:
> `self._alm_name = tmp_json['alm_name']`
> to
> `self._alm_name = tmp_json.get('alm_name')`
> 
> @nsimb
> `__init__.py` row 507:
> `self._optinapp = tmp_json['optInApp']`
> to
> `self._optinapp = tmp_json.get('optInApp')`
> 
> If you could try this and see if the error moves to another row, that would be of great help to me. If this json string handling is fixing the error, then I could release a new beta that is corrected.

Looks like I missed allot (sorry was away) 
Will try to test the new version this afternoon

**Comment by nsimb** [comment]
did install the new version, did however not do the manual changes @onkelfarmor did (since my setup worked fine with the _ in "return self._model_description"
Looks like this ones are working:
sensor.indego_batt_voltage
sensor.indego_battery
sensor.indego_lawn_mowed
sensor.indego_mower_state
sensor.indego_mowing_mode
sensor.indego_runtime_total

Alerts, update are not available, same as @onkelfarmor missing

not sure if its might to this  in sensor.py:
#    mower_alert_sensor_name = CONF_MOWER_NAME + ' mower alert'
#    add_devices([IndegoAlertSensor(mower_alert_sensor_name)])

**Comment by jm-73** [comment]
The alerts was commented out because I was experiencing strange errors with the API-calls. I think it is related to UTF8-encoding as my alerts are in swedish (containing swedish special characters).

I have worked on new code to implement the new PYPI API. I will not add any more time on the beta as it takes time from pushing towards the release of a better coded component.

**Comment by jm-73** [comment]
Closed when 0.5 is released.
